File: src/main/java/oscar/visual/VisualArrow.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.visual;

import javax.swing.*;
import java.awt.*;
import java.awt.geom.AffineTransform;

import scala.Tuple2;
import oscar.visual.shapes.*;

/**
 * 
 * @author Pierre Schaus
 *
 */
public class VisualArrow extends VisualLine{
	
	
	
	Polygon arrowHead = new Polygon(); 
	
	public VisualArrow(VisualDrawing d,double xorig, double yorig, double xdest, double ydest,int dim) {
		super(d,xorig,yorig,xdest,ydest);
		
		 
		arrowHead.addPoint( 0,dim);
		arrowHead.addPoint( -dim, -dim);
		arrowHead.addPoint( dim,-dim);

	}
	
    @Override
    public void draw(Graphics2D g) {
    	super.draw(g);
    	drawArrowHead(g);
    }
    
    private void drawArrowHead(Graphics2D g2d) { 
    	
    	AffineTransform tx = new AffineTransform();
        tx.setToIdentity();
        double angle = Math.atan2(line().y2-line().y1, line().x2-line().x1);
        tx.translate(line().x2, line().y2);
        tx.rotate((angle-Math.PI/2d));
        g2d.fill(tx.createTransformedShape(arrowHead));
    }
    

	public static void main(String[] args) {


		try {
			VisualFrame f = new VisualFrame("toto");
			VisualDrawing d = new VisualDrawing(false);
			JInternalFrame inf = f.createFrame("Drawing");
			inf.add(d);
			f.pack();

			VisualArrow arrow = new VisualArrow(d, 50, 50, 100, 50,5);
			
			Thread.sleep(1000);
			
			arrow.dest_$eq(new Tuple2(100.0, 100.0));
		} catch (InterruptedException e) {
			e.printStackTrace();
		}

	}
	
	
}

File: oscar-visual/src/main/scala/oscar/visual/shapes/VisualArrow.scala
package oscar.visual.shapes

import oscar.visual.VisualDrawing
import java.awt.geom.Line2D
import java.awt.Polygon
import java.awt.Graphics2D
import java.awt.geom.AffineTransform
import oscar.visual.VisualFrame
import java.awt.Color
import java.awt.BasicStroke

class VisualArrow(d: VisualDrawing, s: Line2D.Double, dim: Int) extends VisualLine(d, s) {

  fill = true
  
  private val arrowHead: Polygon = new Polygon
  arrowHead.addPoint(0, dim)
  arrowHead.addPoint(-dim, -dim)
  arrowHead.addPoint(dim, -dim)

  override def draw(g2d: Graphics2D): Unit = {    
    val oldFill = fill
    fill = false
    super.draw(g2d)
    fill = oldFill
    // Draw the arrow head
    if (visible) {
      drawArrowHead(g2d)
    }
  }

  private def drawArrowHead(g2d: Graphics2D): Unit = {
    // Position of the head
    val transform = new AffineTransform()
    val (x1, y1) = dest
    val (x2, y2) = orig
    val angle: Double = math.atan2(y2 - y1, x2 - x1)
    transform.translate(x2, y2)
    transform.rotate(angle - math.Pi / 2)
    val transformedHead = transform.createTransformedShape(arrowHead)
    // Draw setup
    val oldStroke = g2d.getStroke()
    if (fill) {
      g2d.setColor(innerCol)
      g2d.fill(transformedHead)
    }
    if (border) {
      g2d.setStroke(plainStroke)
      g2d.setColor(outerCol)
      g2d.draw(transformedHead)
      g2d.setStroke(oldStroke)
    }
  }
}

object VisualArrow {

  def apply(drawing: VisualDrawing, xOrig: Double, yOrig: Double, xDest: Double, yDest: Double, dim: Int = 5): VisualArrow = {
    new VisualArrow(drawing, new Line2D.Double(xOrig, yOrig, xDest, yDest), dim)
  }
}

object VisualArrowTest extends App {

  val frame = VisualFrame("Example")
  val drawing = VisualDrawing(false)
  val inFrame = frame.createFrame("Arrow")
  inFrame.add(drawing)
  frame.pack()

  val arrow = VisualArrow(drawing, 50, 50, 100, 50, 5)
  arrow.innerCol = Color.red
  Thread.sleep(1000)
  arrow.dest = (100, 100)
  Thread.sleep(1000)
  arrow.borderWidth = 3
  Thread.sleep(1000)
  arrow.dashed = true
  Thread.sleep(1000)
  arrow.border = false
  Thread.sleep(1000)
  arrow.border = true
  arrow.fill = false
}